### PR TITLE
Feature/optgroup support

### PR DESCRIPTION
**Pull request by manuakasam**
References #370 

@gianarb this would be the first set of feature with which adding is not any danger at all I feel.

However there's still one thing that I wouldn't see as "that awesome".

When your data-structure "optgroup" is empty or null it will create an optgroup that has no title. While this is somewhat OK it would likely be more valid to put those empty optgroup entries within NO optgroup block at all.

A further alternative is to provide yet another option key `optgroup_default` where you could define what happens to empty optgroup elements. I.e. `'optgroup_default' => 'Others'` would name the empty group "Others".

The above functionality is already implemented in my branch. You can check it out here:
https://github.com/manuakasam/DoctrineModule/blob/feature/optgroup-default-container/src/DoctrineModule/Form/Element/Proxy.php#L617-L630

The downside to this is that this is quite a lot of custom code. While fully functional you may consider this too custom and not for a wide enough audience? Just gimme your thouhgt. If it's considered OK, ill write tests for those as well and hit up a follow up PR after this.

### CODE REVIEW

**Comment by gianarb** [comment]
In my opinion is OK! :)

File: src/DoctrineModule/Form/Element/Proxy.php [hunk 1]
```diff
@@ -297,14 +311,31 @@ public function getLabelGenerator()
...
+    {
+        return $this->optgroupIdentifier;
+    }
+
+    /**
+     * @param string $optgroupIdentifier
+     */
+    public function setOptgroupIdentifier($optgroupIdentifier)
+    {
+        $this->optgroupIdentifier = $optgroupIdentifier;
```
Comment (danizord) [inline src/DoctrineModule/Form/Element/Proxy.php]: Missing cast to string

File: src/DoctrineModule/Form/Element/Proxy.php [hunk 2]
```diff
@@ -297,14 +311,31 @@ public function getLabelGenerator()
...
      *
-     * @param  boolean         $method
+     * @param  boolean $method
+     *
      * @return Proxy
      */
     public function setIsMethod($method)
     {
-        $this->isMethod = (bool) $method;
+        $this->isMethod = (bool)$method;
```
Comment (danizord) [inline src/DoctrineModule/Form/Element/Proxy.php]: Revert this one

File: src/DoctrineModule/Form/Element/Proxy.php [hunk 3]
```diff
@@ -406,7 +442,7 @@ protected function loadObjects()
             return;
         }
 
-        $findMethod = (array) $this->getFindMethod();
+        $findMethod = (array)$this->getFindMethod();
```
Comment (danizord) [inline src/DoctrineModule/Form/Element/Proxy.php]: Revert

File: src/DoctrineModule/Form/Element/Proxy.php [hunk 4]
```diff
@@ -519,7 +557,7 @@ protected function loadValueOptions()
                     );
                 }
 
-                $label = (string) $object;
+                $label = (string)$object;
```
Comment (danizord) [inline src/DoctrineModule/Form/Element/Proxy.php]: Revert

File: tests/DoctrineModuleTest/Form/Element/TestAsset/FormObject.php [hunk 5]
```diff
@@ -70,7 +75,8 @@ public function __toString()
      */
     public function setId($id)
     {
-        $this->id = (int) $id;
+        $this->id = (int)$id;
```
Comment (danizord) [inline tests/DoctrineModuleTest/Form/Element/TestAsset/FormObject.php]: Revert

File: tests/DoctrineModuleTest/Form/Element/TestAsset/FormObject.php [hunk 6]
```diff
@@ -81,7 +87,8 @@ public function setId($id)
      */
     public function setEmail($email)
     {
-        $this->email = (string) $email;
+        $this->email = (string)$email;
```
Comment (danizord) [inline tests/DoctrineModuleTest/Form/Element/TestAsset/FormObject.php]: Revert

File: tests/DoctrineModuleTest/Form/Element/TestAsset/FormObject.php [hunk 7]
```diff
@@ -100,7 +107,8 @@ public function getEmail()
      */
     public function setPassword($password)
     {
-        $this->password = (string) $password;
+        $this->password = (string)$password;
```
Comment (danizord) [inline tests/DoctrineModuleTest/Form/Element/TestAsset/FormObject.php]: Revert

File: tests/DoctrineModuleTest/Form/Element/TestAsset/FormObject.php [hunk 8]
```diff
@@ -119,7 +127,8 @@ public function getPassword()
      */
     public function setUsername($username)
     {
-        $this->username = (string) $username;
+        $this->username = (string)$username;
```
Comment (danizord) [inline tests/DoctrineModuleTest/Form/Element/TestAsset/FormObject.php]: Revert

File: tests/DoctrineModuleTest/Form/Element/TestAsset/FormObject.php [hunk 9]
```diff
@@ -138,7 +147,8 @@ public function getUsername()
      */
     public function setFirstname($firstname)
     {
-        $this->firstname = (string) $firstname;
+        $this->firstname = (string)$firstname;
```
Comment (danizord) [inline tests/DoctrineModuleTest/Form/Element/TestAsset/FormObject.php]: Revert

File: tests/DoctrineModuleTest/Form/Element/TestAsset/FormObject.php [hunk 10]
```diff
@@ -157,7 +167,8 @@ public function getFirstname()
      */
     public function setSurname($surname)
     {
-        $this->surname = (string) $surname;
+        $this->surname = (string)$surname;
```
Comment (danizord) [inline tests/DoctrineModuleTest/Form/Element/TestAsset/FormObject.php]: Revert

File: src/DoctrineModule/Form/Element/Proxy.php [hunk 11]
```diff
@@ -529,17 +567,37 @@ protected function loadValueOptions()
...
+                $optgroupGetter = 'get' . ucfirst($optgroupIdentifier);
+
+                if (!is_callable(array($object, $optgroupGetter))) {
+                    throw new RuntimeException(
+                        sprintf('Method "%s::%s" is not callable', $this->targetClass, $optgroupGetter)
+                    );
+                }
+
+                $optgroup = $object->{$optgroupGetter}();
+
```
Comment (danizord) [inline src/DoctrineModule/Form/Element/Proxy.php]: Should probably check if `$optgroup` is string, or cast it to string.

**Comment by danizord** [comment]
@manuakasam awesome! Thanks for this feature :+1:

File: src/DoctrineModule/Form/Element/Proxy.php [hunk 12]
```diff
@@ -85,6 +85,11 @@ class Proxy implements ObjectManagerAwareInterface
      */
     protected $emptyItemLabel = '';
 
+    /**
+     * @var string
```
Comment (danizord) [inline src/DoctrineModule/Form/Element/Proxy.php]: string|null

File: src/DoctrineModule/Form/Element/Proxy.php [hunk 13]
```diff
@@ -297,14 +311,31 @@ public function getLabelGenerator()
     }
 
     /**
+     * @return string
```
Comment (danizord) [inline src/DoctrineModule/Form/Element/Proxy.php]: `string|null`

File: src/DoctrineModule/Form/Element/Proxy.php (hunk 1 above)
Comment (manuakasam) [inline src/DoctrineModule/Form/Element/Proxy.php]: Regarding this and related comments: why?
Or more polite: where is the coding standard definition for this. Taking a look at the Proxy class alone sometimes it's done way A sometimes B.

File: src/DoctrineModule/Form/Element/Proxy.php (hunk 2 above)
Comment (manuakasam) [inline src/DoctrineModule/Form/Element/Proxy.php]: I'll certainly revert those, however I'd like to have another opinion on this as the casting isn't defined in either PSR nor the doctrine standards. And a quick irc poll was in favor to this writing style i used :P But as I've mentioned, if there's another say I'll revert.

File: src/DoctrineModule/Form/Element/Proxy.php (hunk 12 above)
Comment (manuakasam) [inline src/DoctrineModule/Form/Element/Proxy.php]: Doesn't go in hand with the rest of the class :\

File: src/DoctrineModule/Form/Element/Proxy.php (hunk 11 above)
Comment (manuakasam) [inline src/DoctrineModule/Form/Element/Proxy.php]: agreed on this place.

**Comment by manuakasam** [comment]
Added the default group handling. Still missing the documentation for that though. Also need more feedback on the proper code-styling of `(string)$foo` vs. `(string) $foo` as well as codeblocks for `@var string` vs. `@var string|null`.

File: src/DoctrineModule/Form/Element/Proxy.php (hunk 1 above)
Comment (danizord) [inline src/DoctrineModule/Form/Element/Proxy.php]: Because you expect `$this->optgroupIdentifier` to be always a string, correct? So there are two ways to ensure that you always set a string to `$this->optgroupIdentifier`:
* Use `is_string()` check and throw exception for invalid arguments (more strict, but more hard to implement and test);
* Cast it to string (easy and painless! It does not solve the problem at all, but it's better than nothing :D).

File: src/DoctrineModule/Form/Element/Proxy.php (hunk 2 above)
Comment (danizord) [inline src/DoctrineModule/Form/Element/Proxy.php]: I'm not sure what's the best way to be honest.

There's a [fixer](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/1.9/Symfony/CS/Fixer/Symfony/SpacesCastFixer.php) in PHPCSFixer to ensure that there's a single space between cast and variable, but I don't know if there's a fixer to ensure the opposite.

Anyway, I think CS fixes should be done in a separate PR.

File: src/DoctrineModule/Form/Element/Proxy.php (hunk 12 above)
Comment (danizord) [inline src/DoctrineModule/Form/Element/Proxy.php]: If constructor does not force `$this->optgroupIdentifier` to be string, then it can be `null` as well. Not sure if you need to follow the "rest of the class" standard here.

File: src/DoctrineModule/Form/Element/Proxy.php (hunk 2 above)
Comment (manuakasam) [inline src/DoctrineModule/Form/Element/Proxy.php]: Alright agreed then. I'll revert all the casting cs then.

File: src/DoctrineModule/Form/Element/Proxy.php (hunk 1 above)
Comment (manuakasam) [inline src/DoctrineModule/Form/Element/Proxy.php]: Alright, I'll still go with the is_string method. More clear and throws the correct exceptions. Stops people from passing a DateTime object as ident :P

File: src/DoctrineModule/Form/Element/Proxy.php (hunk 3 above)
Comment (gianarb) [inline src/DoctrineModule/Form/Element/Proxy.php]: @manuakasam your last commit (CS_FIX) doesn't fix this problem

**Comment by gianarb** [comment]
Ok! :) 
ping @danizord for the last check :P 
Thanks!

**Comment by danizord** [comment]
:100: :ok_hand:

**Comment by gianarb** [comment]
Thanks @manuakasam for this feature and @danizord for the help! :)

**Comment by manuakasam** [comment]
Thanks for accepting it ;) more to come...
